### ISSUE TITLE
multigpu gradient averaging if grad_op is IndexedSlices; race condition in apply_gradients?

### ISSUE THREAD

**Issue by MInner** [tensorflow/tensorflow#5252]
If one follows the [`cifar10_multi_gpu_train.py`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py) example (build two towers that share weights, feed data independently, collect gradients, average them, apply then) and passes gradients to the [`average_gradients()`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/models/image/cifar10/cifar10_multi_gpu_train.py#L110) function, but graph uses `tf.gather` consequently, the resulting gradient op will be `IndexedSlices` that has many `-1` values in `indices` field (from my experience, not sure about why this happens), so when tensorflow will attempt to apply `tensorflow.python.ops.gradients._IndexedSlicesToTensor` to convert it to Tensor (to apply `tf.expand_dims` on it), it will fail on [`unsorted_segment_sum`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/gradients.py#L92) , because `-1`s are indeed not in the indexes range. I have explicitly printed `w_grad_op.indices` and they looked like: 

```
[0 1 2 3 6 4 5 6 7 8 9 13 12 11 15 16 ... 230 231 -1 -1 ... -1 280 -1 -1 265 -1 -1 ... -1]
```

\- so the `average_gradients()` indeed fails on these examples (people have been reporting similar problems [[1]](http://stackoverflow.com/questions/37074077/fail-to-average-indexedslices-on-porting-ptb-word-lm-py-to-multi-gpu-towers), [[2]](http://stackoverflow.com/questions/39017896/tensorflow-how-to-average-several-indexedslicesvalue) .. 2-3 more around the web).

One way of dealing with (solution 1) it is just applying gradients consequentially (with lower learning rate) instead of averaging them first:

```
 train_op = tf.group(*[gd.apply_gradients(grad) for grad in tower_grads])
```

\-  this works great and scales nicely, but (in my case) led to some weird convergence issues (eventually leading to `nan`s popping here and there), possibly because of some sort of race conditions (?), while applying gradients (`gd.apply_gradients()` is probably not very atomic). One possible workaround (solution 2) might be to use `tf.control_dependencies`:

```
def rigid_op_sequence(op_lambdas):
    ## equivalent to:
    ## op = op1()
    ## with.control_dependencies([op]):
    ##    op = op2()
    ##    with.control_dependencies([op]):
    ##        op = op3()
    ##        ...
    ## return op

    with ExitStack() as stack:
        for op_func in op_lambdas:
            op = op_func()
            context = tf.control_dependencies([op])
            stack.enter_context(context)
    return op

grad_update_lambdas = [(lambda: gd.apply_gradients(grad)) for grad in tower_grads]
train_op = rigid_op_sequence(grad_update_lambdas)
```

however this solution seems to somewhat significantly degenerate performance, because of that "blocking" (GPU load dropped from ~90% to ~60%).

Several fundamental questions:
1. What is officially recommended way of dealing with this "multigpu IndexedSlices" case? (could not find any; people keep asking)
2. Does (solution 1) indeed might experience race condition or tensorflow handles it somehow and my convergences issues were not related to that? (and we all can just use solution 1)

**Comment by drpngx** [comment]
@mrry who might have better insight

**Comment by aselle** [comment]
@ebrevdo, could you comment on this?

**Comment by MInner** [comment]
UPD: after little more testing on relatively large language model, it turned out that even though `nvidia-smi` shows lower gpu utilization when I'm using "blocking" solution (2), it is scales great (on 2 gpus the speedup is 1.83x) and performs actually even better (10% faster) then non-blocking one (1).

**Comment by ebrevdo** [comment]
@MInner having -1s in the IndexedSlices object output of gather sounds like a bug.  do you know if gather() is being called with -1s for the indices input?

**Comment by girving** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by TomorrowIsAnOtherDay** [comment]
I have the same problem when trying to implement a multigpu version model with embedding layer.

What is officially recommended way of dealing with this "multigpu IndexedSlices" case? (could not find any; people keep asking)

**Comment by vanpersie32** [comment]
@TomorrowIsAnOtherDay ,hi do you have any idea of average multigpu indexedslices?

**Comment by JerrikEph** [comment]
I have the same problem, I am wondering if this is solved in new versions of tensorflow? or is there any solution for averaging IndexedSlices.

**Comment by guptapriya** [comment]
@yuefengz can maybe shed some light on this.

**Comment by JerrikEph** [comment]
Given that `indices` in `IndexedSlices` from different gpus are the same, this should work.
```python
def average_gradients(tower_grads):
  average_grads = []
  for grad_and_vars in zip(*tower_grads):

    if isinstance(grad_and_vars[0][0], tf.IndexedSlices):
        grad_indices = grad_and_vars[0][0].indices
        grad_shape = grad_and_vars[0][0].dense_shape
        grad_value = tf.add_n([g.values for g, _ in grad_and_vars]) / float(len(tower_grads))
        grad = tf.IndexedSlices(values=grad_value, indices=grad_indices, dense_shape=grad_shape)

    else:
        grad = tf.add_n([g for g, _ in grad_and_vars]) / float(len(tower_grads))

    v = grad_and_vars[0][1]
    grad_and_var = (grad, v)
    average_grads.append(grad_and_var)

  return average_grads
```

It manually extract `values` and `indices` from `IndexedSlices` and then average `values`, and re-encapsulate `averaged_value` and `indices` into an IndexedSlices Object. 
Do note that it is only applicable when the indices from different `IndexedSlices` are identical.
Hope it helps, somebody should really implement basic operations for `IndexedSlices`.

**Comment by liyi193328** [comment]
Any ideas? want to know the solutions

**Comment by guptapriya** [comment]
tf.distribute.MirroredStrategy handles aggregating IndexedSlices already afaik. How are you doing multi GPU training?
